### PR TITLE
MM-21898: Part 2. Add opentracing

### PR DESCRIPTION
**Pull request by reflog**
#### Summary
This PR continues the work in https://github.com/mattermost/mattermost-server/pull/13840 by implementing a opentracing wrapper layer on top of App/Store.
The following changes have been made:
1. `EnableOpenTracing `setting added (false by default)
2. If `EnableOpenTracing` is true - Jaeger client is instantiated
3. If `EnableOpenTracing `is true - Each HTTP request wraps App and Store in OpenTracing layers
4. Context is added to App and Store and is passed down the layer chain to create OpenTracing 'spans'
5. **By default no method parameters are reported to opentracing** to avoid leaking sensitive information. Only the name of the method is reported
6. To enable parameter reporting, a comment should be added above Store interface methods or App struct methods to list the params that need to be traced. For examples see `ExecuteCommand`, `CreateCommandPost `methods in App and `GetFlaggedPostsForTeam` in PostStore.
  
#### Ticket Link
Fixes https://mattermost.atlassian.net/browse/MM-21898

#### Screenshot
After tracing, Jaeger can now display:
image

### CODE REVIEW

File: app/layer_generators/main.go [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+			return strings.Join(paramsWithType, ", ")
+		},
+	}
+
+	t := template.Must(template.New("opentracing_layer.go.tmpl").Funcs(myFuncs).ParseFiles(templateFile))
+	err := t.Execute(out, metadata)
+	if err != nil {
+		panic(err)
+	}
+	return out.String()
```
Comment (agnivade) [inline app/layer_generators/main.go]: It seems like we are returning string and then again converting to []byte slice to the next function. What do you think about returning a byte slice instead, so that double allocations are avoided ?

File: app/layer_generators/main.go [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+	"path"
+	"strings"
+	"text/template"
+
+	"golang.org/x/tools/imports"
+)
+
+var reserved = []string{"AcceptLanguage", "AccountMigration", "Cluster", "Compliance", "Context", "DataRetention", "Elasticsearch", "HTTPService", "ImageProxy", "IpAddress", "Ldap", "Log", "MessageExport", "Metrics", "Notification", "NotificationsLog", "Path", "RequestId", "Saml", "Session", "SetIpAddress", "SetRequestId", "SetSession", "SetStore", "SetT", "Srv", "Store", "T", "Timezones", "UserAgent", "SetUserAgent", "SetAcceptLanguage", "SetPath", "SetContext", "SetServer", "GetT"}
+
+var outputFile string
```
Comment (agnivade) [inline app/layer_generators/main.go]: Declare this in a var group.

File: app/layer_generators/main.go [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+var inputFile string
+var outputFileTemplate string
+
+const OPEN_TRACING_PARAMS_MARKER = "@openTracingParams"
+
+func main() {
+	flag.StringVar(&inputFile, "in", path.Join("..", "app_iface.go"), "App interface file")
```
Comment (agnivade) [inline app/layer_generators/main.go]: Do the flag declarations in an init function.

File: app/layer_generators/main.go [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+	flag.StringVar(&outputFile, "out", path.Join("..", "opentracing_layer.go"), "Output file")
+	flag.StringVar(&outputFileTemplate, "template", "opentracing_layer.go.tmpl", "Output template file")
+	flag.Parse()
+	BuildOpenTracingAppLayer()
+}
+
+func BuildOpenTracingAppLayer() {
```
Comment (agnivade) [inline app/layer_generators/main.go]: I don't see a need for this to be in a separate function. There are no tests for this. Let's unexport this and put this inside main itself.

File: app/layer_generators/main.go [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+				}
+			}
+		}
+
+		return true
+	})
+	return metadata
+}
+
+func GenerateLayer(name, templateFile string) string {
```
Comment (agnivade) [inline app/layer_generators/main.go]: Same as above. Let's unexport this.

File: app/layer_generators/main.go [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+		panic(err)
+	}
+
+	err = ioutil.WriteFile(outputFile, formattedCode, 0644)
+	if err != nil {
+		panic(err)
+	}
+}
+
+type Param struct {
```
Comment (agnivade) [inline app/layer_generators/main.go]: unexport of all these structs and functions.

File: app/layer_generators/main.go [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+	flag.StringVar(&outputFileTemplate, "template", "opentracing_layer.go.tmpl", "Output template file")
+	flag.Parse()
+	BuildOpenTracingAppLayer()
+}
+
+func BuildOpenTracingAppLayer() {
+	code := GenerateLayer("OpenTracingAppLayer", outputFileTemplate)
+	formattedCode, err := imports.Process(outputFile, []byte(code), &imports.Options{Comments: true})
+	if err != nil {
+		panic(err)
```
Comment (agnivade) [inline app/layer_generators/main.go]: Let's avoid panics here and everywhere else. We can bubble up the error with appropriate context, and call a log.Fatal in the main if we need to exit immediately. Else a simple log.Println and return.

File: app/layer_generators/main.go [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+		},
+		"genResultsVars": func(results []string) string {
+			vars := []string{}
+			for idx := range results {
+				vars = append(vars, fmt.Sprintf("resultVar%d", idx))
+			}
+			return strings.Join(vars, ", ")
+		},
+		"errorToBoolean": func(results []string) string {
+			for idx, typeName := range results {
```
Comment (agnivade) [inline app/layer_generators/main.go]: s/idx/i

File: app/layer_generators/main.go [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+	out := bytes.NewBufferString("")
+	metadata := ExtractStoreMetadata()
+	metadata.Name = name
+
+	myFuncs := template.FuncMap{
+		"joinResults": func(results []string) string {
+			return strings.Join(results, ", ")
+		},
+		"joinResultsForSignature": func(results []string) string {
+			if len(results) == 0 {
```
Comment (agnivade) [inline app/layer_generators/main.go]: Let's use a switch case on `len(results)`.

File: app/layer_generators/main.go [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+				return ""
+			}
+			if len(results) == 1 {
+				return strings.Join(results, ", ")
+			}
+			return fmt.Sprintf("(%s)", strings.Join(results, ", "))
+		},
+		"genResultsVars": func(results []string) string {
+			vars := []string{}
```
Comment (agnivade) [inline app/layer_generators/main.go]: Pre-allocate a slice with a cap of len(results).

File: app/layer_generators/main.go [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+			for idx := range results {
```
Comment (agnivade) [inline app/layer_generators/main.go]: s/idx/i

File: app/layer_generators/main.go [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+				vars = append(vars, fmt.Sprintf("resultVar%d", idx))
+			}
+			return strings.Join(vars, ", ")
+		},
+		"errorToBoolean": func(results []string) string {
+			for idx, typeName := range results {
+				if typeName == "*model.AppError" {
```
Comment (agnivade) [inline app/layer_generators/main.go]: Let's have a constant for "*model.AppError" and reuse that.

File: app/layer_generators/main.go [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+		"errorPresent": func(results []string) bool {
+			for _, typeName := range results {
+				if typeName == "*model.AppError" {
+					return true
+				}
+			}
+			return false
+		},
+		"errorVar": func(results []string) string {
+			for idx, typeName := range results {
```
Comment (agnivade) [inline app/layer_generators/main.go]: s/idx/i

File: store/layer_generators/main.go [hunk 14]
```diff
@@ -92,10 +115,20 @@ func ExtractStoreMetadata() StoreMetadata {
 					if _, ok := topLevelFunctions[methodName]; ok {
 						params := []Param{}
 						results := []string{}
-
+						paramsToTrace := map[string]bool{}
 						ast.Inspect(method.Type, func(expr ast.Node) bool {
```
Comment (agnivade) [inline store/layer_generators/main.go]: This ast.Inspect handler for x.Name.Name == "store" and the else if condition below looks suspiciously similar. Any chance we could refactor these to a separate function ?

File: store/layer_generators/main.go [hunk 15]
```diff
@@ -92,10 +115,20 @@ func ExtractStoreMetadata() StoreMetadata {
...
+										s := comment.Text
+										if idx := strings.Index(s, OPEN_TRACING_PARAMS_MARKER); idx != -1 {
+											for _, p := range strings.Split(s[idx+len(OPEN_TRACING_PARAMS_MARKER):], ",") {
+												paramsToTrace[strings.TrimSpace(p)] = true
+											}
+										}
+									}
+								}
 								if e.Params != nil {
 									for _, param := range e.Params.List {
```
Comment (agnivade) [inline store/layer_generators/main.go]: I was thinking of an edge-case. What if someone changes a param name and then forgets to update the opentracing param comment ? Can we add a check here after collecting the `paramsToTrace` and `params`, where we just iterate `paramsToTrace` and confirm that every element is present in `params` ? If not, then raise an error.

File: store/layer_generators/main.go [hunk 16]
```diff
@@ -17,8 +17,15 @@ import (
...
 
+const OPEN_TRACING_PARAMS_MARKER = "@openTracingParams"
+
 func main() {
-	code := GenerateTimerLayer()
+	BuildTimerLayer()
+	BuildOpenTracingLayer()
+}
+
+func BuildTimerLayer() {
```
Comment (agnivade) [inline store/layer_generators/main.go]: Similar comments as the other main function.
- Unexport functions and structs.
- Remove panics.

File: app/app_iface.go [hunk 17]
```diff
@@ -36,16 +36,263 @@ import (
 
 // AppIface is extracted from App struct and contains all it's exported methods. It's provided to allow partial interface passing and app layers creation.
 type AppIface interface {
```
[the remark this answers is not in the document]
Comment (reflog) [reply app/app_iface.go]: Let's move this to another issue (together with removal of Get* getters), since this is not really related to this PR.

File: app/app_iface.go [hunk 18]
```diff
@@ -36,16 +36,263 @@ import (
...
+	FillInPostProps(post *model.Post, channel *model.Channel) *model.AppError
+	// FilterNonGroupChannelMembers returns the subset of the given user IDs of the users who are not members of groups
+	// associated to the channel excluding bots
+	FilterNonGroupChannelMembers(userIds []string, channel *model.Channel) ([]string, error)
+	// FilterNonGroupTeamMembers returns the subset of the given user IDs of the users who are not members of groups
+	// associated to the team excluding bots.
+	FilterNonGroupTeamMembers(userIds []string, team *model.Team) ([]string, error)
+	// GetAllLdapGroupsPage retrieves all LDAP groups under the configured base DN using the default or configured group
+	// filter.
+	GetAllLdapGroupsPage(page int, perPage int, opts model.LdapGroupSearchOpts) ([]*model.Group, int, *model.AppError)
```
[the remark this answers is not in the document]
Comment (reflog) [reply app/app_iface.go]: Same as above.

File: store/layer_generators/timer_layer.go.tmpl [hunk 19]
```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+// Code generated by "make store-layers"
+// DO NOT EDIT
+
+package store
+
+import (
+	timemodule "time"
```
[the remark this answers is not in the document]
Comment (reflog) [reply store/layer_generators/timer_layer.go.tmpl]: A parameter to some of the functions is called 'time'. So aliasing is needed.

**Comment by reflog** [comment]
@agnivade - Thanks for an amazing review. I've addressed the comments

File: app/app_iface.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (agnivade) [reply app/app_iface.go]: @reflog -  Do you want to file the issue, or should I ?

[the remark this answers is not in the document]
Comment (reflog) [reply app/app_iface.go]: @agnivade I'll do it.

File: app/layer_generators/main.go [hunk 20]
```diff
@@ -0,0 +1,248 @@
...
+		}
+		if !found {
+			log.Fatalf("Unable to find a parameter called '%s' (method '%s') that is mentioned in the '%s' comment. Maybe it was renamed?", paramName, method.Names[0].Name, OPEN_TRACING_PARAMS_MARKER)
+		}
+	}
+	return methodData{Params: params, Results: results, ParamsToTrace: paramsToTrace}
+}
+
+func extractStoreMetadata() (*storeMetadata, error) {
+
```
Comment (mgdelacroix) [inline app/layer_generators/main.go]: Unnecessary whitespace

File: app/layer_generators/opentracing_layer.go.tmpl [hunk 21]
```diff
@@ -0,0 +1,218 @@
...
+	span, newCtx := tracing.StartSpanWithParentByContext(a.ctx, "app.{{$index}}")
+
+	a.ctx = newCtx
+	a.app.Srv().Store.SetContext(newCtx)
+	defer func() { 
+		a.app.Srv().Store.SetContext(origCtx)
+		a.ctx = origCtx 
+	}()
+	{{range $paramIdx, $param := $element.Params}}
+		{{if index $element.ParamsToTrace $param.Name }}
```
Comment (mgdelacroix) [inline app/layer_generators/opentracing_layer.go.tmpl]: nit:
```suggestion
		{{if index $element.ParamsToTrace $param.Name}}
```

File: app/layer_generators/opentracing_layer.go.tmpl [hunk 22]
```diff
@@ -0,0 +1,218 @@
...
+			span.SetTag("{{$param.Name}}", {{$param.Name}})
+		{{end}}	
+	{{end}}
+	defer span.Finish()
+	{{- if $element.Results | len | eq 0}}
+		a.app.{{$index}}({{$element.Params | joinParams}})
+	{{ else }}
```
Comment (mgdelacroix) [inline app/layer_generators/opentracing_layer.go.tmpl]: nit:
```suggestion
	{{else}}
```

File: app/layer_generators/opentracing_layer.go.tmpl [hunk 23]
```diff
@@ -0,0 +1,218 @@
...
+		{{$element.Results | genResultsVars}} := a.app.{{$index}}({{$element.Params | joinParams}})
+		{{if $element.Results | errorPresent }}
```
Comment (mgdelacroix) [inline app/layer_generators/opentracing_layer.go.tmpl]: nit:
```suggestion
		{{if $element.Results | errorPresent}}
```

File: app/layer_generators/opentracing_layer.go.tmpl [hunk 24]
```diff
@@ -0,0 +1,218 @@
...
+			if {{$element.Results | errorVar}} != nil {
+				span.LogFields(spanlog.Error({{$element.Results | errorVar}}))
+				ext.Error.Set(span, true)
+			}
+		{{end}}		
+		return {{$element.Results | genResultsVars -}}
+	{{ end}}}
```
Comment (mgdelacroix) [inline app/layer_generators/opentracing_layer.go.tmpl]: nit:
```suggestion
	{{end}}}
```

File: app/server.go [hunk 25]
```diff
@@ -374,6 +385,11 @@ func (s *Server) Shutdown() error {
 
 	s.RunOldAppShutdown()
 
+	if s.tracer != nil {
+		err := s.tracer.Close()
+		mlog.Error("Unable to cleanly shutdown opentracing client", mlog.Err(err))
```
Comment (mgdelacroix) [inline app/server.go]: logging the error should only happen if an error actually happens:

```go
if err := s.tracer.Close(); if err != nil {
	mlog.Error("Unable to cleanly shutdown opentracing client", mlog.Err(err))
}
```

File: store/layer_generators/opentracing_layer.go.tmpl [hunk 26]
```diff
@@ -0,0 +1,82 @@
...
+	}()
+	{{range $paramName, $param := $element.Params}}
+		{{if index $element.ParamsToTrace $param.Name }}
+			span.SetTag("{{$param.Name}}", {{$param.Name}})
+		{{end}}
+	{{end}}
+	defer span.Finish()
+	{{- if $element.Results | len | eq 0}}
+	s.{{$substoreName}}Store.{{$index}}({{$element.Params | joinParams}})
+	{{ else }}
```
Comment (mgdelacroix) [inline store/layer_generators/opentracing_layer.go.tmpl]: nit:
```suggestion
	{{else}}
```

File: store/layer_generators/opentracing_layer.go.tmpl [hunk 27]
```diff
@@ -0,0 +1,82 @@
...
+{{range $index, $element := $substore.Methods}}
+func (s *{{$.Name}}{{$substoreName}}Store) {{$index}}({{$element.Params | joinParamsWithType}}) {{$element.Results | joinResultsForSignature}} {
+	origCtx := s.Root.Store.Context()
+	span, newCtx := tracing.StartSpanWithParentByContext(s.Root.Store.Context(), "{{$substoreName}}Store.{{$index}}")
+	s.Root.Store.SetContext(newCtx)
+	defer func(){
+		s.Root.Store.SetContext(origCtx)
+	}()
+	{{range $paramName, $param := $element.Params}}
+		{{if index $element.ParamsToTrace $param.Name }}
```
Comment (mgdelacroix) [inline store/layer_generators/opentracing_layer.go.tmpl]: nit:
```suggestion
		{{if index $element.ParamsToTrace $param.Name}}
```

File: store/layer_generators/opentracing_layer.go.tmpl [hunk 28]
```diff
@@ -0,0 +1,82 @@
...
+			span.SetTag("{{$param.Name}}", {{$param.Name}})
+		{{end}}
+	{{end}}
+	defer span.Finish()
+	{{- if $element.Results | len | eq 0}}
+	s.{{$substoreName}}Store.{{$index}}({{$element.Params | joinParams}})
+	{{ else }}
+		{{$element.Results | genResultsVars}} := s.{{$substoreName}}Store.{{$index}}({{$element.Params | joinParams}})
+		{{- if $element.Results | errorPresent }}
```
Comment (mgdelacroix) [inline store/layer_generators/opentracing_layer.go.tmpl]: nit:
```suggestion
		{{- if $element.Results | errorPresent}}
```

File: store/layer_generators/opentracing_layer.go.tmpl [hunk 29]
```diff
@@ -0,0 +1,82 @@
...
+			if {{$element.Results | errorVar}} != nil {
+				span.LogFields(spanlog.Error({{$element.Results | errorVar}}))
+				ext.Error.Set(span, true)
+			}
+		{{end}}	
+		return {{ $element.Results | genResultsVars -}}
+	{{ end }}
```
Comment (mgdelacroix) [inline store/layer_generators/opentracing_layer.go.tmpl]: nit:
```suggestion
	{{end}}
```

File: store/layer_generators/opentracing_layer.go.tmpl [hunk 30]
```diff
@@ -0,0 +1,82 @@
...
+}
+{{end}}
+{{end}}
+
+{{range $index, $element := .Methods}}
+func (s *{{$.Name}}) {{$index}}({{$element.Params | joinParamsWithType}}) {{$element.Results | joinResultsForSignature}} {
+	{{if $element.Results | len | eq 0}}s.Store.{{$index}}({{$element.Params | joinParams}})
+	{{ else }}return s.Store.{{$index}}({{$element.Params | joinParams}})
+	{{ end}}}
```
Comment (mgdelacroix) [inline store/layer_generators/opentracing_layer.go.tmpl]: nit:
```suggestion
	{{else}}return s.Store.{{$index}}({{$element.Params | joinParams}})
	{{end}}
}
```

File: store/layer_generators/timer_layer.go.tmpl [hunk 31]
```diff
@@ -0,0 +1,78 @@
...
+
+{{end}}
+
+{{range $substoreName, $substore := .SubStores}}
+{{range $index, $element := $substore.Methods}}
+func (s *{{$.Name}}{{$substoreName}}Store) {{$index}}({{$element.Params | joinParamsWithType}}) {{$element.Results | joinResultsForSignature}} {
+	start := timemodule.Now()
+	{{if $element.Results | len | eq 0}}
+	s.{{$substoreName}}Store.{{$index}}({{$element.Params | joinParams}})
+	{{ else }}
```
Comment (mgdelacroix) [inline store/layer_generators/timer_layer.go.tmpl]: nit:
```suggestion
	{{else}}
```

File: store/layer_generators/timer_layer.go.tmpl [hunk 32]
```diff
@@ -0,0 +1,78 @@
...
+	{{$element.Results | genResultsVars}} := s.{{$substoreName}}Store.{{$index}}({{$element.Params | joinParams}})
+	{{ end }}
```
Comment (mgdelacroix) [inline store/layer_generators/timer_layer.go.tmpl]: nit:
```suggestion
	{{end}}
```

File: store/layer_generators/timer_layer.go.tmpl [hunk 33]
```diff
@@ -0,0 +1,78 @@
...
+	{{- end }}
+}
+{{end}}
+{{end}}
+
+{{range $index, $element := .Methods}}
+func (s *{{$.Name}}) {{$index}}({{$element.Params | joinParamsWithType}}) {{$element.Results | joinResultsForSignature}} {
+	{{if $element.Results | len | eq 0}}s.Store.{{$index}}({{$element.Params | joinParams}})
+	{{ else }}return s.Store.{{$index}}({{$element.Params | joinParams}})
+	{{ end}}}
```
Comment (mgdelacroix) [inline store/layer_generators/timer_layer.go.tmpl]: nit:
```suggestion
	{{else}}return s.Store.{{$index}}({{$element.Params | joinParams}})
	{{end}}
}
```

**Comment by cpatulea** [comment]
I am trying to investigate some performance issue and looking for tracing support.

For posterity, opentracing support was removed in January 2025 (commit f1acdce42c0ea4238086cbbd85b030e55800abdd).
